core/src/main/java/io/crate/types/TableType.java

```diff
@@ -0,0 +1,171 @@
+ * May be used in an IN or ANY operation, e.g. select 1 where 1 = ANY(Select 1)
+ */
+public class TableType extends DataType<Collection> implements CollectionType, Streamer<Object> {
+
+    public static final int ID = 102;
+    private DataType<?> valueType;
+
+    public TableType(DataType<?> innerType) {
+        this.valueType = innerType;
+    }
+
+    TableType() {}
+
+    @Override
+    public int id() {
+        return ID;
+    }
+
+    @Override
+    public String getName() {
+        return valueType.getName() + "_table";
+    }
+
+    @Override
+    public Streamer<?> streamer() {
+        return this;
+    }
+
+    @Override
+    public Collection value(Object value) {
+        if (value == null) {
+            return null;
+        }
+        if (value instanceof Collection) {
+            return (Collection) value;
+        }
+        return Collections.singleton(valueType.value(value));
+    }
+
+    @Override
+    public boolean isConvertableTo(DataType other) {
+        return this.valueType.isConvertableTo(other) ||
+               (other != null && other instanceof CollectionType &&
+                ((CollectionType) other).innerType().isConvertableTo(this.valueType));
+    }
+
+    @Override
+    public int compareValueTo(Collection val1, Collection val2) {
+        if (val1 == val2 || (val1 != null && val1.equals(val2))) {
+            return 0;
+        }
+        return -1;
+    }
+
+    @Override
+    public int compareTo(Object o) {
+        if (!(o instanceof TableType)) {
+            return -1;
+        }
+        return Integer.compare(valueType.id(), ((TableType) o).valueType.id());
+    }
+
+    @Override
+    public void readFrom(StreamInput in) throws IOException {
+        valueType = DataTypes.fromStream(in);
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        DataTypes.toStream(valueType, out);
+    }
+
+    @Override
+    public Collection readValueFrom(StreamInput in) throws IOException {
+        if (valueType.id() == SetType.ID || valueType.id() == ArrayType.ID) {
+            return (Collection) valueType.streamer().readValueFrom(in);
+        }
+        int size = in.readVInt();
+        switch (size) {
+            case 0:
+                return Collections.EMPTY_LIST;
+            case 1:
+                return Collections.singletonList(valueType.streamer().readValueFrom(in));
+            default:
+                List<Object> list = new ArrayList<>(size);
+                for (int i = 0; i < size; i++) {
+                    list.add(valueType.streamer().readValueFrom(in));
+                }
+                return list;
+        }
+    }
+
+    @Override
+    public void writeValueTo(StreamOutput out, Object value) throws IOException {
+        if (value == null) {
+            out.writeVInt(0);
+            return;
+        }
+        if (valueType.id() == SetType.ID || valueType.id() == ArrayType.ID) {
```

Comment:
This is to support the new `TableType` with the existing `SetType` or `ArrayType`. Since they all return `Collection` values, we want to avoid nesting them.

We still support `Select 1 where 1 = ANY(Select collect_set(id) from t) which now wraps the `SetType` into a `TableType`. It's not totally obvious, we should add a comment.